src/QMCWaveFunctions/WaveFunctionComponent.h

```diff
@@ -267,37 +308,127 @@ struct WaveFunctionComponent : public QMCTraits
+  /** If a move for iat-th particle is rejected, restore to the content.
    * @param iat index of the particle whose new position was proposed
+   *
+   * Ye: hopefully we can gradually move away from restore
    */
   virtual void restore(int iat) = 0;
 
+  /** If a move for iat-th particle on some walkers in a batch is rejected, restore their contents
+   *  Note that all the lists only include rejected walkers.
+   * @param WFC_list the list of WaveFunctionComponent pointers of the same component in a walker batch
+   * @param iat index of the particle whose new position was proposed
+   *
+   * Ye: hopefully we can gradually move away from restore
+   */
+  virtual void mw_restore(const std::vector<WaveFunctionComponent*>& WFC_list, int iat)
+  {
+    //#pragma omp parallel for
+    for (int iw = 0; iw < WFC_list.size(); iw++)
+      WFC_list[iw]->restore(iat);
+  }
+
   /** evaluate the ratio of the new to old WaveFunctionComponent value
-   *@param P the active ParticleSet
-   *@param iat the index of a particle
-   *@return \f$ \psi( \{ {\bf R}^{'} \} )/ \psi( \{ {\bf R}^{'}\})\f$
+   * @param P the active ParticleSet
+   * @param iat the index of a particle
+   * @return \f$ \psi( \{ {\bf R}^{'} \} )/ \psi( \{ {\bf R}\})\f$
    *
-   *Specialized for particle-by-particle move.
+   * Specialized for particle-by-particle move
    */
   virtual ValueType ratio(ParticleSet& P, int iat) = 0;
 
+  /** compute the ratio of the new to old WaveFunctionComponent value of multiple walkers
+   * @param WFC_list the list of WaveFunctionComponent pointers of the same component in a walker batch
+   * @param P_list the list of ParticleSet pointers in a walker batch
+   * @param iat particle index
+   * @param ratios the list of WF ratios of a walker batch, \f$ \Psi( \{ {\bf R}^{'} \} )/ \Psi( \{ {\bf R}\})\f$
+   */
+  virtual void mw_calcRatio(const std::vector<WaveFunctionComponent*>& WFC_list,
+                            const std::vector<ParticleSet*>& P_list,
+                            int iat,
+                            std::vector<PsiValueType>& ratios)
+  {
+    #pragma omp parallel for
+    for (int iw = 0; iw < WFC_list.size(); iw++)
+      ratios[iw] = WFC_list[iw]->ratio(*P_list[iw], iat);
+  }
+
   /** For particle-by-particle move. Requests space in the buffer
    *  based on the data type sizes of the objects in this class.
    * @param P particle set
    * @param buf Anonymous storage
    */
   virtual void registerData(ParticleSet& P, WFBufferType& buf) = 0;
 
+  /** For particle-by-particle move. Requests space in the buffer
+   *  based on the data type sizes of the objects in this class.
+   * @param WFC_list the list of WaveFunctionComponent pointers of the same component in a walker batch
+   * @param P_list the list of ParticleSet pointers in a walker batch
+   * @param buf_list Anonymous storage
+   */
+  virtual void mw_registerData(const std::vector<WaveFunctionComponent*>& WFC_list,
```

Comment:
The whole buffer model seems like it should go ASAP.  I looks very hacky and it adds a set of semantics to constructing/moving/reverting objects completely outside of normal C++ idiom.

Reply:
I believe many people needs this pattern that a big chunk of memory is reserved and each object owns part of it when using. When communicating or copy the memory, it is used as a whole. I do believe we can do better C++ but his pattern doesn't go away. The other option is serialize the data by copying in and out a buffer. I consider this pattern worse because the packing/unpack interfaces are still need, more memory used, more time spent on moving memory.

Reply:
to be discussed further probably post unified driver.